2.30.8 8.2 linstor fixes staging use make available

rp- opened the pull request:
Instead of using diskless_on_remaing, only place a resource on demand on a node.
If a cluster would ever exceed 20 nodes, having diskless on all of them
might get problematic, as all of them are part of the quorum voting.


I'm not sure what base branch I should use to do the PR against,
so just tell me if I should rebase it against some other branch.

Wescoeur committed:
feat(drivers): add LinstorSR driver

Some important points:

- linstor.KV must use an identifier name that starts with a letter (so it uses a "sr-" prefix).

- Encrypted VDI are supported with key_hash attribute (not tested, experimental).

- When a new LINSTOR volume is created on a host (via snapshot or create), the remaining diskless
devices are not necessarily created on other hosts. So if a resource definition exists without
local device path, we ask it to LINSTOR. Wait 5s for symlink creation when a new volume
is created => 5s is is purely arbitrary, but this guarantees that we do not try to access the
volume if the symlink has not yet been created by the udev rule.

- Can change the provisioning using the device config 'provisioning' param.

- We can only increase volume size (See: https://github.com/LINBIT/linstor-server/issues/66),
it would be great if we could shrink volumes to limit the space used by the snapshots.

- Inflate/Deflate can only be executed on the master host, a linstor-manager plugin is present
to do this from slaves. The same plugin is used to open LINSTOR ports + start controller.

- Use a `total_allocated_volume_size` method to have a good idea of the reserved memory
Why? Because `physical_free_size` is computed using the LVM used size, in the case of thick provisioning it's ok,
but when thin provisioning is choosen LVM returns only the allocated size using the used block count. So this method
solves this problem, it takes the fixed virtual volume size of each node to compute the required size to store the
volume data.

- Call vhd-util on remote hosts using the linstor-manager when necessary, i.e. vhd-util is called to get vhd info,
the DRBD device can be in use (and unusable by external processes), so we must use the local LVM device that
contains the DRBD data or a remote disk if the DRBD device is diskless.

- If a DRBD device is in use when vhdutil.getVHDInfo is called, we must have no
errors. So a LinstorVhdUtil wrapper is now used to bypass DRBD layer when
VDIs are loaded.

- Refresh PhyLink when unpause in called on DRBD devices:
We must always recreate the symlink to ensure we have
the right info. Why? Because if the volume UUID is changed in
LINSTOR the symlink is not directly updated. When live leaf
coalesce is executed we have these steps:
"A" -> "OLD_A"
"B" -> "A"
Without symlink update the previous "A" path is reused instead of
"B" path. Note: "A", "B" and "OLD_A" are UUIDs.

- Since linstor python modules are not present on every XCP-ng host,
module imports are protected by try.. except... blocks.

- Provide a linstor-monitor daemon to check master changes

Wescoeur committed:
feat(drivers): add CephFS, GlusterFS and XFS drivers

Wescoeur committed:
feat(drivers): add ZFS driver to avoid losing VDI metadata (xcp-ng/xcp#401)

stormi committed:
Update xs-sm.service's description for XCP-ng

This was a patch added to the sm RPM git repo before we had this
forked git repo for sm in the xcp-ng github organisation.

stormi committed:
Add TrueNAS multipath config

This was a patch added to the sm RPM git repo before we had this
forked git repo for sm in the xcp-ng github organisation.

stormi committed:
Re-add the ext4 driver for users who need to transition

The driver is needed to transition to the ext driver.
Users who upgrade from XCP-ng <= 8.0 need a working driver so that they
can move the VMs out of the ext4 SR and delete the SR.

Not keeping that driver would force such users to upgrade to 8.1 first,
convert their SR, then upgrade to a higher version.

However, like in XCP-ng 8.1, the driver will refuse any new ext4 SR
creation.

Wescoeur committed:
feat(tests): add unit tests concerning ZFS (close xcp-ng/xcp#425)

- Check if "create" doesn't succeed without zfs packages
- Check if "scan" failed if the path is not mounted (not a ZFS mountpoint)

Wescoeur committed:
fix(LinstorSR): repair volumes only if an exclusive command is executed

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): Improve LINSTOR performance

Details:
- vdi_attach and vdi_detach are now exclusive
- lock volumes on slaves (when vdi_xxx command is used) and avoid release if a timeout is reached
- load all VDIs only when necessary, so only if it exists at least a journal entry or if sr_scan/sr_attach is executed
- use a __slots__ attr in LinstorVolumeManager to increase performance
- use a cache directly in LinstorVolumeManager to reduce network request count with LINSTOR
- try to always use the same LINSTOR KV object to limit netwok usage
- use a cache to avoid a new JSON parsing when all VDIs are loaded in LinstorSR
- limit request count when LINSTOR storage pool info is fetched using a fetch interval
- avoid race condition in cleanup: check if a volume is locked in a slave or not before modify it
- ...

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): robustify scan to avoid losing VDIs if function is called outside module

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): display a correctly readable size for the user

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(linstor-monitord): scan all LINSTOR SRs every 12 minutes to update allocated size stats

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): call correctly method in _locked_load when vdi_attach_from_config is executed

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): integrate minidrbdcluster daemon

Now, we can:
- Start a controller on any node
- Share the LINSTOR volume list using a specific volume "xcp-persistent-database"
- Use the HA with "xcp-persistent-ha-statefile" and "xcp-persistent-redo-log" volumes
- Create the nodes automatically during SR creation

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

xandrus committed:
Added SM Driver for MooseFS

Co-authored-by: Piotr Robert Konopelko <piotr.konopelko@moosefs.pro>
Signed-off-by: Aleksander Wieliczko <aleksander.wieliczko@moosefs.pro>
Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): ensure heartbeat and redo_log VDIs are not diskless

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

benjamreis committed:
If no NFS ACLs provided, assume everyone:

Some QNAP devices do not provide ACL when fetching NFS mounts.
In this case the assumed ACL should be: "*".

This commit fixes the crash when attempting to access the non existing ACL.
Relevant issues:
- https://github.com/xapi-project/sm/issues/511
- https://github.com/xcp-ng/xcp/issues/113

Wescoeur committed:
feat(LinstorSR): protect sr commands to avoid forgetting LINSTOR volumes when master satellite is down

Steps to reproduce:

- Ensure the linstor satellite is not running on the master host, otherwise stop it
- Then restart the controller on the right host where the LINSTOR database is mounted
- Run st_attach command => All volumes will be forgotten

To avoid this, it's possible to restart the satellite on the master before the sr_attach command.
Also it's funny to see you can start and stop the satellite juste before the sr_attach, and the volumes will not be removed.

Explanations:

In theory this bug is impossible because during the sr_attach execution, an exception is thrown
(so sr_scan should not be executed) BUT there is a piece of code that is executed
in SRCommand.py when sr_attach is called:

```python
try:
    return sr.attach(sr_uuid)
finally:
    if is_master:
        sr.after_master_attach(sr_uuid)
```

The exception is not immediately forwarded because a finally block must be executed before.
And what is the implementation of after_master_attach?

```python
def after_master_attach(self, uuid):
    """Perform actions required after attaching on the pool master
    Return:
    None
    """
    self.scan(uuid)
```

Oh! Of course, a scan is always executed after a attach... What's the purpose of a scan if we can't
execute correctly an attach command before? I don't know, but it's probably error-prone like this context.
When scan is called, we suppose the SR is attached and we have all VDIs loaded but it's not the case
because an exception has been thrown.

To solve this problem we forbid the execution of the scan if the attach failed.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorJournaler): ensure uri is not None during linstor.KV creation

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): add an option to disable auto-quorum on volume DB + fix doc

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorVolumeManager): add a workaround to create properly SR with thin LVM

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): add optional ips parameter

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): add a helper `log_drbd_erofs` to trace EROFS errno code with DRBD resources + check EROFS error

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): try to restart the services again if there is a failure in linstor-manager

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

MarkSymsCtx committed:
backport of ccd121cc248d79b749a63d4ad099e6d5f4b8b588: CA-354692: check for device parameter in create/probe calls

Signed-off-by: Mark Syms <mark.syms@citrix.com>
Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): robustify linstor-manager to never include from plugins path

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): prevent starting controller during fail in linstor manager destroy method

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorVolumeManager): increase peer slots limit (support 31 connections to a DRBD)

- Also, create diskless devices when db is created

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorVolumeManager): add a fallback to find controller uri (when len(hosts) >= 4)

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(var-lib-linstor.mount): ensure we always mount database with RW flags

Sometimes systemd fallback to read only FS if the volume can't be mounted, we must
forbid that. It's probably a DRBD error.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorVolumeManager): add a fallback to find node name (when len(hosts) >= 4)

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): explain on which host, plugins commands are executed

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): create diskless path if necessary during VDI loading

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Avoid usage of `umount` in `ISOSR` when `legacy_mode` is used

`umount` should not be called when `legacy_mode` is enabled, otherwise a mounted dir
used during SR creation is unmounted at the end of the `create` call (and also
when a PBD is unplugged) in `detach` block.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Fix timeout_call: alarm must be reset in case of success

Otherwise the SIGALRM signal can be emitted after the execution
of the given user function.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
timeout_call returns the result of user function now

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): find controller when XAPI unreachable (XHA)

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): use IPs instead of hostnames in NBD server

Without this patch we can't use XCP-ng hosts configured with static IPS.
Only servers with DHCP enabled can access the HTTP server(s) otherwise.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorVolumeManager): ensure we always use IPs in _get_controller_uri

Otherwise if a hostname is returned, we can't use it if the XCP-ng pool
is configurer using static IPs instead of DHCP.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(linstor-manager): add methods to add remove/host from LINSTOR SR

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorVolumeManager): support SR creation with diskless nodes

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): add a config var to disable HTTP/NBD servers

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSr): ensure LVM group is activated during SR.attach/create

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(linstor-manager): add method to create LinstorSR + to list/destroy DRBD volumes

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): always set vdi_path in generate_config

If the volume of a generated config is not related to HTTP/NBD
and if we already have a path to the resource, the VDI path is never
written to the config. So the config can't be used to attach the VDI...

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): use HTTP/NBD instead of DRBD directly with heartbeat VDI

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(minidrbdcluster): supports new properties like `force-io-failures`

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
MooseFS SR uses now UUID subdirs for each SR

A sm-config boolean param `subdir` is available to configure where to store the VHDs:
- In a subdir with the SR UUID, the new behavior
- In the root directory of the MooseFS SR

By default, new SRs are created with `subdir` = True.
Existing SRs  are not modified and continue to use the folder that was given at
SR creation, directly, without looking for a subdirectory.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): enabled/disable minidrbcluster with fixed order

Ensure we disable minidrbdcluster during SR destruction on all hosts
with a deterministic execution to ensure linstor-controller is never restarted
on another host. It was possible before this patch because the host of the minidrbcluster
that was running the controller could be stopped before the others. Now the primary service
is stopped last.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstor-manager): change linstor satellite start behavior

Ensure we don't have an invalid cache used by a satellite:
- We found an issue with a new added disk which used a volume group name
  formerly involved by another disk. To avoid this kind of problem, we
  always restart the satellite.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Fix is_open call for LinstorSR

1. Ensure LinstorSR driver is imported in `_is_open` definition to register it in the driver list.
Otherwise this function always fails with a SRUnknownType exception.

2. Fetch the dconf of the target SR to retrieve VDI path, i.e. we can't use fake params like
other drivers, we must have a real LINSTOR connection to read in the DB the volume location.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvhdutil): fix boolean params of `check` call

`ignoreMissingFooter` and `fast` must be string types to be used with XAPI plugin API.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Fix is_open call for many drivers (#25)

Ensure all shared drivers are imported in `_is_open` definition to register
them in the driver list. Otherwise this function always fails with a SRUnknownType exception.

Also, we must add two fake mandatory parameters to make MooseFS happy: `masterhost` and `rootpath`.
Same for CephFS with: `serverpath`. (NFS driver is directly patched to ensure there is no usage of
the `serverpath` param because its value is equal to None.)

`location` param is required to use ZFS, to be more precise, in the parent class: `FileSR`.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
Remove SR_CACHING capability for many SR types (#24)

SR_CACHING offers the capacity to use IntelliCache, but this
feature is only available using NFS SR.

For more details, the implementation of `_setup_cache` in blktap2.py
uses only an instance of NFSFileVDI for the shared target.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(linstor-manager): robustify exec_create_sr

- Use lvm.py XCP-ng xapi plugins instead of lvm (old name)
- Check arguments to create the SR
- Fix param types given to SR.create
- lsof_resource use verbose output if there is a lock or problem
- Remove useless `force` param on remove_host

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(cleanup): print LINSTOR VDI UUID if error during info loading (not SR UUID)

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(cleanup): raise and dump DRBD openers in case of bad coalesce

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(linstorvhdutil): trace DRBD openers in case of EROFS errors

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvolumemanager): compute correctly size in allocated_volume_size

Remove replication count in computation.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): use DRBD openers instead of lsof to log in blktap2

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorJournaler): reset namespace when `get` is called

Otherwise, we can be in the wrong namespace and the key to find will be inaccessible.

Wescoeur committed:
fix(linstorvhdutil): fix coalesce with VM running under specific scenario:

When a VM is running, we can't coalesce without this patch with a long chain
of VHDs because a parent can be in use on another host, and so a EROFS
can be emitted by vhd-util.

So to fix this problem we run vhd-util on the remote host instead of the master
in case of failure in the cleanup algorithm.

Impacted vhd-util functions: coalesce, getParent, repair.

Wescoeur committed:
feat(LinstorSR): support cProfile to trace calls when a command is executed

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvolumemanager): `_get_volumes_info` doesn't raise with offline host

Ensure this method doesn't raise an exception when a host is offline.
Otherwise we can't use properly the HA when a host is unreachable and it's a
problem to restart VMs on a valid host.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvolumemanager): remove double prefix on kv group name

- Before this patch, when the kv store was created/accessed, a double "xcp-sr-" prefix was used.
- This change is not compatible with existing LINSTOR SR instances!

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): add linstor-kv-dump helper to print kv store

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): disable VHD key hash usage to limit exec time

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(minidrbdcluster): ensure SIGINT is handled correctly

This patch is here to make sure no LINSTOR controller survives when
systemd asks to minidrbdcluster to stop with `SIGINT`.

- Remove `os.system`, it's totally unsafe, all signals are ignored with it.
- Use `subprocess.Popen` instead and catch correctly signal exceptions, it works
  because `wait` call doesn't hide the signals.
- Ensure `SIGINT` is only sent to the main process, not to the subprocesses.
- Ensure `SIGKILL` is NEVER sent to minidrbdcluster.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(minidrbdcluster): stop resource services at startup

- Ensure all services are stopped when minidrbcluster is started.
- Clean code to parse only once the systemd unit string.
- Log unhandled exceptions.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(linstor-manager): add new `healthCheck` function to monitor pool (#26)

Print a JSON output to monitor state of LINSTOR SRs:
  - Display nodes, storage pool and resources
  - Print human readable warns and errors

Usage example:

```
xe host-call-plugin host-uuid=c96ec4dd-28ac-4df4-b73c-4371bd202728 plugin=linstor-manager fn=healthCheck args:groupName=linstor_group
{
  "errors": [],
  "warns": [],
  "controller-uri": "linstor://172.16.210.14",
  "storage-pools": {
    "r620-s1": [
      {
        "free-size": 999125155840,
        "storage-pool-name": "xcp-sr-linstor_group",
        "capacity": 1000203091968,
        "uuid": "994a5c45-ba52-4f17-8e46-74c7dec0a1e7"
      }
    ],
    "r620-s3": [
      {
        "free-size": 999125155840,
        "storage-pool-name": "xcp-sr-linstor_group",
        "capacity": 1000203091968,
        "uuid": "ad78adad-a9f6-4513-9f96-e8eb8fe716dc"
      }
    ],
    "r620-s2": [
      {
        "free-size": 999125155840,
        "storage-pool-name": "xcp-sr-linstor_group",
        "capacity": 1000203091968,
        "uuid": "f76048f9-8821-484b-9a51-670a49df7a6e"
      }
    ]
  },
  "nodes": {
    "r620-s1": "ONLINE",
    "r620-s3": "ONLINE",
    "r620-s2": "ONLINE"
  },
  "resources": {
    "xcp-persistent-database": {
      "r620-s1": {
        "tie-breaker": false,
        "in-use": true,
        "volumes": [
          {
            "storage-pool-name": "xcp-sr-linstor_group",
            "uuid": "1a436f23-eb81-4a8f-8ab6-de317282b5d5",
            "device-path": "/dev/drbd1000",
            "number": 0,
            "disk-state": "UpToDate",
            "allocated-size": 1077936128,
            "usable-size": 1073741824
          }
        ],
        "diskful": true
      },
      "r620-s3": {
        "tie-breaker": false,
        "in-use": false,
        "volumes": [
          {
            "storage-pool-name": "xcp-sr-linstor_group",
            "uuid": "31a05bd1-20b6-471a-86b9-bbcdccfaab96",
            "device-path": "/dev/drbd1000",
            "number": 0,
            "disk-state": "UpToDate",
            "allocated-size": 1077936128,
            "usable-size": 1073741824
          }
        ],
        "diskful": true
      },
      "r620-s2": {
        "tie-breaker": false,
        "in-use": false,
        "volumes": [
          {
            "storage-pool-name": "xcp-sr-linstor_group",
            "uuid": "0420f252-9762-4063-bdd4-732e40373ffb",
            "device-path": "/dev/drbd1000",
            "number": 0,
            "disk-state": "UpToDate",
            "allocated-size": 1077936128,
            "usable-size": 1073741824
          }
        ],
        "diskful": true
      }
    }
  }
}
```

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): fix xha conf parsing => return host ip, not the UUID

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): start correctly HA servers (HTTP/NBD) after reboot

Use a timeout call after a reboot to get a XAPI session because
it can be impossible to initialize the connection at startup
or if the current host has been ejected (in this case the call can
block indefinitely).

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvolumemanager): use an array to store diskful volumes info

Otherwise the `is_diskful` attr only reflects the info of one host
after a call to `get_volume_info`... And we therefore lose this
information concerning the others.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(linstorvolumemanager): support snaps when a host is offline

- Don't create diskless volumes during clone, delay it.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvolumemanager): support offline hosts when plugins are called

- Robustify plugin calls
- Add a timeout on session getter

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvolumemanager): define _base_group_name member at SR creation

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(linstorvhdutil): modify logic of local vhdutil calls

- Always log openers when we can't call vhdutil locally
- Remove "raise" lines after a local fail, always retry command calls on a remote

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvolumemanager): robustify failed snapshots

- Ensure we can always rename a failed snap, so we must check if
  we have metadata in the KV-store. Otherwise an error is triggered
  because we are trying to copy a None object.
- If we can't delete a volume, rename it with a DELETED_ suffix,
  it's mandatory in specific cases to rollback snapshot transactions.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvolumemanager): use a namespace for volumes

- This change is not compatible with existing LINSTOR SR instances!

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(linstor-kv-dump): rename to linstor-kv-tool + add remove volume helpers

Wescoeur committed:
fix(LinstorSR): handle correctly localhost during start/stop of minidrbdcluster

Otherwise another controller can be started during `xe sr-destroy` call.
And in this case we can get an exception during the umount/remount of /var/lib/linstor
to destroy the persistent database because a controller uses it.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(cleanup.py): call repair on another host when EROFS is returned (DRBD)

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): avoid introduction of DELETED volumes

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(linstor-kv-tool): remove-all-volumes supports journals now

Not yet supported for remove-volume, not sure about the consequences
for the moment.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvhdutil): due to bad refactoring, check call was broken

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(linstorvhdutil): ensure we use VHD parent to find host where to coalesce

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(linstorvolumemanager): force DRBD demote after failed volume creation/clone

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvhdutil): ensure we retry creation in all situations

Without this patch, a basic resource creation is never restarted
after a failure. The classic situation is when a DRBD error is raised like
`Failed to adjust DRBD resource`. This problem is rare but it exists.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvhdutil): don't retry local vhdutil call when EROFS is detected

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(fork-log-daemon): ignore SIGTERM

Without this patch, the output logs of the fork-log-daemon child
are never displayed when SIGTERM is sent to the PGID.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): wait for http-disk-server startup

Avoid a race condition with NBD server.
We must be sure the HTTP server is reachable before the NBD server execution,
otherwise the HA activation may fail.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): handle inflate + resize actions correctly

- Ensure LINSTOR set the expected new volume size when inflate is executed,
  otherwise we log and we use the returned size.
- Repair deflate calls in case of journal entry, cache is not usable.
- Logs VHD and DRBD volume size in create/resize methods.
- Ensure resize is only executed on master.
- Use utilisation size instead of capacity in journal entries.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstor-manager): add a static iptables rule for DRBD volumes

Using the XAPI iptables firewall may drop DRBD packets when the connection
tracking subsystem runs out of entries temporarily.

Instead, use a static rule completely independent of the connection tracking
module, it allow packets to pass even when the connection tracking table
ran full temporarily.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): sync with last http-nbd-transfer version

- Increase auto promote timeout of heartbeat VDI to reduce CPU usage
- Modify server regexes
- Force device size parameter of NBD servers

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): don't check VDI metadata while listing VDIs if it's deleted

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): don't check metadata when destroying snap in undo_clone

Remove useless check in the snap rollback helper when there is an error
during the second `_create_snapshot` call (when VDI.clone command is executed).

There is no reason to verify the metadata, this code is
present in the LVM driver, but is useless here.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvhdutil): handle correctly generic exceptions in _raise_openers_exception

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(minidrbdcluster): robustify to unmount correctly LINSTOR DB

There is a small delay during which the database may not be unmounted
because there are still processes using it. So must retry in this case.
It's caused by the termination of the LINSTOR controller.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(minidrbdcluster): handle correctly KeyboardInterrupt with systemd units

It's necessary to always add systemd services in the running list before
trying to start a service, because if a KeyboardInterrupt is sent, we can have
a running LINSTOR controller not present in the list, and then we can no longer
unmount /var/lib/linstor because the controller is never stopped...

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
feat(LinstorSR): use drbd-reactor instead of minidrbdcluster

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): ensure vhdutil calls are correctly executed on pools with > 3 hosts

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): replace bad param in detach_thin impl

To get the physical size, the volume UUID must be used, not the path.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvolumemanager): remove usage of realpath

Because a diskless DRBD path not always exist, get_volume_name_from_device_path can fail.
It's easy to reproduce using > 4 hosts and with a call to linstorvhdutil.get_vhd_info:
This problem can occur if the parent of a VHD is not on the same machine and if this parent doesn't have a DRBD path locally.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvhdutil): avoid parent path resolution

When many hosts are used (>= 4), we can fail to get
VHD info (with parent option) because the local parent VDI
path can be absent (no DRBD diskless path). So it's necessary
to deactivate parent resolution:
- vhdutil has been patched to support that
- vhdutil returns a relative path now when "-u" option is used

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): create parent path during attach

It's necessary to force DRBD diskless path creation when
a VDI is attached. Otherwise the attach can fail on pool with
>= 4 hosts.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): retry if we can't build volume cache

Otherwise after SR creation, the master PBD can be unplugged.
See: https://xcp-ng.org/forum/post/60726

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstorvolumemanager): reduce peer-slots param to 3

Because we use 3 backing disks at most, it's useless to increase the default linstor limit (8).
Diskless is a resource that does not count in the peer-slots param.

Note: this change is important to reduce the RAM usage, see =>
https://linbit.com/drbd-user-guide/drbd-guide-9_0-en/#s-meta-data-size
Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): attach a valid XAPI session is_open is called

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): ensure we always have a DRBD path to snap

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): remove hosts/ips param

Wescoeur committed:
fix(LinstorSR): compute correctly SR size using pool count

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(blktap2): ensure we can import this module when LINSTOR is not installed

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): ensure volume cache can be recreated

After SR creation we may fail to load volumes with this exception:
"Failed to get usable size of..." and so we can't plug the master PBD.

Regardless of the retry timeout, the only solution to fetch the usable
size of the DB is to recreate the connection to the LINSTOR API.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(linstor-manager): remove dead/useless code in add/remove_host helpers

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

rp- committed:
linstor: don't create diskless on all nodes

Instead of using diskless_on_remaing, only place a resource on demand
on a node.
If a cluster would ever exceed 20 nodes, having diskless on all of them
might get problematic, as all of them are part of the quorum voting.

Signed-off-by: Rene Peinthor <rene.peinthor@linbit.com>

Wescoeur committed:
fix(LinstorSR): Ensure we always have a device path during leaf-coalesce calls

So we must not verify that we have a valid DRBD path in the load step,
it can fail on many hosts, instead we must create a diskless path only
during the real coalesce.

Note: I removed this assert: `assert virtual_size >= volume_size`,
it seems that it's not always true, I suppose the volume size can be
greater than expected due to a bigger allocation in the LVM or DRBD layer.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur commented on drivers/LinstorSR.py:
```diff
@@ -1834,6 +1834,8 @@ def attach(self, sr_uuid, vdi_uuid):
                 'scan SR first to trigger auto-repair'
             )
 
+        self._linstor.make_volume_available(volume_uuid=vdi_uuid)
```
This line is redundant with this piece of code that handles the current VDI + the VHD parents: https://github.com/xcp-ng/sm/pull/34/files#diff-29e9cda22a2d2236948b509b32f46bcdc1148b21ed92f798baa6053f7fdf478dR1877

Wescoeur commented on drivers/linstorvolumemanager.py:
```diff
@@ -785,6 +785,30 @@ def ensure_volume_list_is_not_locked(self, volume_uuids, timeout=None):
         if waiting:
             self._logger('No volume locked now!')
 
+    def make_volume_available(self, volume_uuid):
```
So this new function is probably not useful like I said in my previous message, but we can probable use `resource_make_available` function instead of `resource_create` in `_activate_device_path`. What do you think?

rp- replied:
Ok, I did miss, that a resource_create was "hidden" in the get_device_path.
I used make_available now instead of the resource_create (but they nearly do the same thing).

Wescoeur commented on drivers/LinstorSR.py:
```diff
@@ -1924,6 +1926,8 @@ def detach(self, sr_uuid, vdi_uuid):
                     .format(e)
                 )
 
+        self._linstor.remove_volume_if_diskless(volume_uuid=vdi_uuid)
```
Because this method has only one argument, we can remove the explicit keyword param.

Wescoeur commented on drivers/linstorvolumemanager.py:
```diff
@@ -785,6 +785,30 @@ def ensure_volume_list_is_not_locked(self, volume_uuids, timeout=None):
+
+    def remove_volume_if_diskless(self, volume_uuid):
+        node_name = socket.gethostname()
+        volume_properties = self._get_volume_properties(volume_uuid)
```
Just a missing line before: `self._ensure_volume_exists(volume_uuid)`

Wescoeur commented on drivers/linstorvolumemanager.py:
```diff
@@ -785,6 +785,30 @@ def ensure_volume_list_is_not_locked(self, volume_uuids, timeout=None):
+            raise LinstorVolumeManagerError(
+                'volume `{}` could not be deleted `{}`'
+                .format(volume_uuid, node_name),
+                LinstorVolumeManagerError.ERR_GENERIC
```
You can remove `LinstorVolumeManagerError.ERR_GENERIC`, it's already the default argument.

Wescoeur commented on drivers/linstorvolumemanager.py:
```diff
@@ -785,6 +785,30 @@ def ensure_volume_list_is_not_locked(self, volume_uuids, timeout=None):
+        replies = self._linstor.resource_delete_if_diskless(node_name=node_name, rsc_name=volume_name)
+        if not linstor.Linstor.all_api_responses_no_error(replies):
+            raise LinstorVolumeManagerError(
+                'volume `{}` could not be deleted `{}`'
```
```suggestion
                'Volume `{}` could not be deleted on `{}`'
```

Wescoeur commented:
Don't forget to amend your commit for the DCO with: `git commit --amend  --signoff`.
The destination branch is correct.

benjamreis committed:
Remove `SR_PROBE` from ZFS capabilities (#37)

The probe method is not implemented so we
shouldn't advertise it.

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

Wescoeur committed:
fix(LinstorSR): always use lock.acquire() during attach/detach

We can't use a retry range on the lock because we can trigger a bad situation
in the detach step... When the GC has a lock on the SR and we try to acquire
the same lock in a detach call, and if we can't get this lock after 20 seconds,
the consequences are very bad:
- Many tapdisk instances of the same VDI can be created on two hosts
- The VDI info are not updated correctly
- And this issue is not immediately visible

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

Wescoeur committed:
fix(LinstorSR): mare sure hostnames are unique at SR creation

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>

rp- commented:
@Wescoeur rebased against latest force push

Wescoeur commented:
I will reread next week, there is a missing part to change, one helper in linstorvolumemanager.py:

```python
def _create_volume
```

It has a special param to remove: `no_diskless`, but it requires that I reread carefully our code.

Wescoeur commented:
Hello, sorry for this long delay, I was in vacation for few weeks, and we had some rather urgent subjects juste before, so I'm back. If that's not a problem for you, I'll probably create a new PR this week adding my fixes to your changes.

Wescoeur commented:
@rp- I see a potential problem with the `resource_delete_if_diskless` call after a smapi detach command.
Is it risky to call this helper when a process has a fd on the diskless path?

To explain the situation, in a specific case, the smapi GC can try to open the DRBD diskless path before or during the `resource_delete_if_diskless` call.

rp- commented:
The API call would certainly fail if there are still openers to the drbd device.
I'm not familiar enough with the SM API to know where a better place would be to call it.

Not calling it at all, would probably leave a lot of unused diskless resources, like we have now(with diskless on remaining).

Wescoeur commented:
Ok, so I can try to execute this call in our SR lock to ensure there is no issue with the GC.
There are probably few places where we can remove diskless volumes, but I'm not sure that is always feasible due to few limitations. I'm trying to see what I can do without impacting the existing code too much.

FYI, I created this branch: https://github.com/xcp-ng/sm/commits/ran-diskless-improvements
I'm going to add all my changes related to diskless, I'll open a PR on the repo after that.

Wescoeur commented:
Hi, so I will release a new RPM with few fixes regarding the diskless volumes:
- https://github.com/xcp-ng/sm/commit/836cf724dd53387f206b061763e1f2fe208f6ad7
- https://github.com/xcp-ng/sm/commit/9366dfd1c3a6129f072fb104714eb37fce1e28b4

Diskless can be created during snapshot steps or coalesce. So:
- These volumes are not cleaned on the master host, because it can be really complex to deal with the GC and parallel tasks.
- By default the diskless are removed during VM shutdown (the VHD parents are also impacted).
- That's not perfect but enough for the global cases.
- I tried to handle complicated cases but honestly I don't think it's a good idea to do that in smapi v1.

Wescoeur commented:
Feel free to add your comments. If you don't see any thing to fix we can close this PR/discussion. Thanks again for your patience. :)
